File: Problem_Solving/leetcode/Algorithm/Binary_Search/704_Binary_Seearch/binarySearch.py
# 4가지 방법

# 재귀 풀이
def search1(nums: List[int], target: int) -> int:
    def binary_search(left, right):
        if left <= right:
            mid = left + (right - left) // 2

            if nums[mid] < target:
                return binary_search(mid + 1, right)
            elif nums[mid] > target:
                return binary_search(left, mid - 1)
            else:
                return mid

        else:
            return -1

    return binary_search(0, len(nums) - 1)

# 반복 풀이


def search2(nums: List[int], target: int) -> int:
    left, right = 0, len(nums) - 1
    while left <= right:
        mid = left + (right - left) // 2

        if nums[mid] < target:
            left = mid + 1
        elif nums[mid] > target:
            right = mid - 1
        else:
            return mid
        
    return -1

# 이진 검색 모듈(파이썬은 이진 검색 알고리즘을 지원하는 bisect 모듈이 있다.)
def search3(nums: List[int], target: int) -> int:
    index = bisect.bisect_left(nums, target)
    
    if index < len(nums) and nums[index] == target:
        return index
    else:
        return -1
    
# 이진 검색을 사용하지 않는 index 풀이
def search4(nums: List[int], target: int) -> int:
    # 해당 값의 index를 찾아해는 함수
    try:
        return nums.index(target)
    # 존재하지 않는 값은 Error가 발생하므로 예외처리
    except ValueError:
        return -1

File: Problem_Solving/leetcode/Algorithm/Binary_Search/704_Binary_Seearch/BinarySearch.cpp
/*
    704. Binary Search
        Given an array of integers nums which is sorted in ascending order, 
        and an integer target, write a function to search target in nums. 
        If target exists, then return its index. Otherwise, return -1.

        You must write an algorithm with O(log n) runtime complexity.

    Example 1:
        Input: nums = [-1,0,3,5,9,12], target = 9
        Output: 4
        Explanation: 9 exists in nums and its index is 4

    Example 2:
        Input: nums = [-1,0,3,5,9,12], target = 2
        Output: -1
        Explanation: 2 does not exist in nums so return -1
    
    Constraints:
        1 <= nums.length <= 104
        -104 < nums[i], target < 104
        All the integers in nums are unique.
        nums is sorted in ascending order.
*/
#include <vector>

class Solution {
public:
    int binarySearch(std::vector<int>& nums, int& left, int& right, int& target)
    {
        // 절반씩 범위를 줄여나가며 재귀호출
        if (left <= right)
        {
            // 중간 값을 비교해 분할해나간다.
            int mid = (left + right) / 2;

            if (nums[mid] < target)
                return binarySearch(nums, ++mid, right, target);
            else if (nums[mid] > target)
                return binarySearch(nums, left, --mid, target);
            else
                return mid;
        }
        else
            return -1;
    }
    int search(std::vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;
        return binarySearch(nums, left, right, target);
    }
};